src/python/WMCore/Services/DDM/DDM.py

```diff
@@ -0,0 +1,286 @@
+                continue
+            try:
+                if ddmReq[key] != self[key]:
+                    return False
+            except KeyError:
+                return False
+
+        for key in ddmReq:
+            if key in exclude:
+                continue
+            try:
+                if ddmReq[key] != self[key]:
+                    return False
+            except KeyError:
+                return False
+        return True
+
+
+class DDM(Service):
+    """
+    Class which provides client APIs to the DDM service.
+
+    Usage example:
+        api = 'copy'
+        ddmReq = DDMReqTemplate(api)
+        ddm = DDM()
+        ddm.makeRequest(ddmReq, api)
+    """
+
+    def __init__(self, url=None, logger=None, configDict=None):
+        """
+        configDict:
+            Dictionary with parameters that are passed to the super class.
+        """
+        url = url or "https://dynamo.mit.edu"
+        configDict = configDict or {}
+        configDict.setdefault('endpoint', url)
+        configDict.setdefault('cacheduration', 1)  # in hours
+        configDict.setdefault('accept_type', 'application/json')
+        configDict.setdefault('content_type', 'application/json')
+        configDict['logger'] = logger if logger else logging.getLogger()
+        super(DDM, self).__init__(configDict)
+        self['logger'].debug("Initializing DDM with url: %s", self['endpoint'])
+
+    def _getResult(self, ddmReq, apiUrl, callname=""):
+        """
+        Either fetch data from the cache file or query the data-service
+        :param metricNumber: a number corresponding to the SSB metric
+        :return: a dictionary
+
+        """
+
+        # TODO: to make this cache not per call but per DDM object instance
+        cachedApi = "%s.json" % callname
+
+        # striping the None values from the request before sending it to dynamo
+        ddmReq.strip()
+        msg = "INFO: sending data to DDM: "
+        msg += "API: {}: ddmReq: {}".format(ddmReq.api, pformat(ddmReq))
+        self['logger'].info(msg)
+
+        # The Request class expects this to be a dict, otherwise it tries to
+        # encode the inputdata once again and inserts parasite information and
+        # breaks the call to dynamo.
+        ddmReq = dict(ddmReq)
+        data = self.refreshCache(cachedApi,
+                                 apiUrl,
+                                 inputdata=ddmReq,
+                                 verb='POST')
+        results = data.read()
+        data.close()
+
+        results = json.loads(results)
+        return results
+
+    def makeRequest(self, ddmReq, api=None):
+        """
+        _makeRequest_
+
+        Makes a SINGLE request to DDM
+
+        :ddmReq: Request template an instance of DDMReqTemplate class.
+        :api:    The DDM api to be queried.
+        :return: Dictionary containing the result from DDM
+        """
+
+        # try to figure out the api from the ddmReq template
+        if not api:
```

Comment:
Perhaps we could make the API name a mandatory argument, and then here we validate it against the `ddmReq.api` attribute?

Reply:
Actually exactly the opposite. While using this module during the tests, I figurred out that having the 'api' in the template and fetching it from there, makes things much more easy to use and understandable (as it is done for the aggregates requests). I would rather drop this argument, or if we keep it, we may keep it just for completeness of the interface - but it will only open the door for future errors, because at some point one can make the mistake to mix a template and a call for a different apis.

Reply:
Done. The api is now fetched from the Template only.